feature/experimental/gribi/ate_tests/route_removal_during_failover_test/route_removal_during_failover_test.go

```diff
@@ -0,0 +1,609 @@
+// findSecondaryController finds out primary and secondary controllers
+func findSecondaryController(t *testing.T, dut *ondatra.DUTDevice, controllers []string) (string, string) {
+	t.Helper()
+	var primary, secondary string
+	for _, controller := range controllers {
+		role := gnmi.Get(t, dut, gnmi.OC().Component(controller).RedundantRole().State())
+		t.Logf("Component(controller).RedundantRole().Get(t): %v, Role: %v", controller, role)
+		if role == secondaryController {
+			secondary = controller
+		} else if role == primaryController {
+			primary = controller
+		} else {
+			t.Fatalf("Expected controller %s to be active or standby, got %v", controller, role)
+		}
+	}
+	if secondary == "" || primary == "" {
+		t.Fatalf("Expected non-empty primary and secondary Controller, got primary: %v, secondary: %v", primary, secondary)
+	}
+	t.Logf("Detected primary: %v, secondary: %v", primary, secondary)
+
+	return secondary, primary
+}
+
+// Function to stop traffic
+func stopTraffic(t *testing.T, ate *ondatra.ATEDevice) {
+	t.Logf("Stopping traffic")
+	ate.Traffic().Stop(t)
+}
+
+// switchoverReady is to check if controller is ready for switchover
+func switchoverReady(t *testing.T, dut *ondatra.DUTDevice, controller string) {
+	switchoverReady := gnmi.OC().Component(controller).SwitchoverReady()
+	gnmi.Await(t, dut, switchoverReady.State(), 30*time.Minute, true)
+}
+
+// validateTelemetry validates telemetry sensors
+func validateSwitchoverTelemetry(t *testing.T, dut *ondatra.DUTDevice, primaryAfterSwitch string) {
+	t.Helper()
+	t.Log("Validate OC Switchover time/reason.")
+	primary := gnmi.OC().Component(primaryAfterSwitch)
+	if !gnmi.Lookup(t, dut, primary.LastSwitchoverTime().State()).IsPresent() {
+		t.Errorf("primary.LastSwitchoverTime().Lookup(t).IsPresent(): got false, want true")
+	} else {
+		t.Logf("Found primary.LastSwitchoverTime(): %v", gnmi.Get(t, dut, primary.LastSwitchoverTime().State()))
+	}
+
+	if !gnmi.Lookup(t, dut, primary.LastSwitchoverReason().State()).IsPresent() {
+		t.Errorf("primary.LastSwitchoverReason().Lookup(t).IsPresent(): got false, want true")
+	} else {
+		lastSwitchoverReason := gnmi.Get(t, dut, primary.LastSwitchoverReason().State())
+		t.Logf("Found lastSwitchoverReason.GetDetails(): %v", lastSwitchoverReason.GetDetails())
+		t.Logf("Found lastSwitchoverReason.GetTrigger().String(): %v", lastSwitchoverReason.GetTrigger().String())
+	}
+}
+
+// checkNIHasNEntries uses the Get RPC to validate that the network instance named ni
+// contains want (an integer) entries.
+func checkNIHasNEntries(ctx context.Context, c *fluent.GRIBIClient, ni string, t testing.TB) int {
+	t.Helper()
+	gr, err := c.Get().
+		WithNetworkInstance(ni).
+		WithAFT(fluent.AllAFTs).
+		Send()
+
+	if err != nil {
+		t.Fatalf("Unexpected error from get, got: %v", err)
+	}
+	return len(gr.GetEntry())
+}
+
+// Send traffic and validate traffic.
+func testTraffic(t *testing.T, args testArgs, flow *ondatra.Flow) {
+	sendTraffic(t, args.ate, flow)
```

Comment:
I don't think sendTraffic and stopTraffic pull their weight as separate functions -- I'd just inline them. I'd probably also inline verifyTraffic but could live with that being separate if you feel strongly.

Reply:
Merged traffic functions